### ISSUE TITLE
Configurable Helm Version

### ISSUE THREAD

**Issue by hawksight** [argoproj/argo-cd#3872]
# Summary

Allow passing the version of helm to use for templating in the application crd / cli.

# Motivation

StatefulSets don't allow changing of labels. Helm has a `.Release.Service` value that is not over-ridable on the command line and this is used in most popular charts to set a label: `heritage: <value>`

I have some applications (rabbitmq, cert-manager) that are installed with helm (version3) and I am trying to adopt these intalls with ArgoCD. Unfortunately for rabbitmq chart, it's a stateful set and cannot change the `heritage` label.

Helm v3 cli sets this label with the value `Helm`, but Helm v2 uses the value `Tiller`. ArgoCD chooses the helm binary based on the chart.yaml `apiVersion` field. Most charts however will still use `v1` for backwards compatibility (I assume).

This means ArgoCD will use helm v2 instead of helm v3 to template most charts. As the value of `.Release.Service` is set by the version of helm being run, this means I need to be able to choose the version of helm that ArgoCD uses. Currently this is not possible and the alternatives are not attractive propositions:

- Hosting own versions of the charts which I'd need to pull and compare on each version increment.
- Asking chart maintainers to adopt `v2` in the `apiVersion` field as it will likely not change unless the chart has a feature dependant on helm3 as not to break backwards compatibility. 

See full discussion under #3735 

# Proposal

Add a new field to the Application CRD to allow setting the version of helm the install uses:

```yaml
spec:
  project: default

  source:
    repoURL: https://github.com/argoproj/argocd-example-apps.git
    targetRevision: HEAD
    path: guestbook

    helm:
      version: v3
      parameters:
      - name: "nginx-ingress.controller.service.annotations.external-dns\\.alpha\\.kubernetes\\.io/hostname"
        value: mydomain.example.com
      releaseName: guestbook
      valueFiles:
      - values-prod.yaml
      values: 
```

In the CLI there may be an optional parameter such as:

```
--helm-version string|int      Specify version of helm to template chart with  
```

If unspecified then the value should be set with the current logic of looking up the `apiVersion` field in chart.yaml.

This is unless a global preference could be specified across an instance of ArgoCD. My preference would be to default to the latest version of helm and only specify the version when `v2` is necessary. However this may be specific to my usage scenario. (Consider this part a stretch goal).

**Comment by jannfis** [comment]
Thanks for this enhancement proposal and sorry for the quite late reaction. I actually do think that this feature would be useful and we should implement it. We have a similar feature for `kustomize`, that goes even a little further by letting the user specify concrete versions of the tool use - many Kustomize versions have behaviour that other versions don't have, and users have worked around those in their manifests. 

Would you think that this make sense for Helm as well, or would it be sufficient to just be able to separate between Helm v2 and v3?

**Comment by hawksight** [comment]
Sorry @jannfis for not getting back to you on this. Honestly I don't use v2 helm and whenever I use a tool I try to stay with the latest version. Im assuming the official images have latest or close to as the binaries you install.

I think if people want a very specific helm version that they should make their own image and overwrite the helm installation to their desired versions. I think that's easy enough to do.

Unfortunately I cannot see any reference to this issue (#3872) or the fix #4111 being included in the 1.7.0 release yesterday:
https://github.com/argoproj/argo-cd/releases/tag/v1.7.0

I don't suppose anyone can say when this will be included? so I can upgrade our argocd instance to try it out.
